File: haskell/greeter/src/greeter.hs
import Data.Char (isSpace)
import System.IO (hFlush, stdout)

getUserName :: IO String
getUserName = do
  putStr "Enter your name: "
  hFlush stdout

  userinput <- getLine
  let name = dropWhile isSpace userinput
  let trimmedName = reverse $ dropWhile isSpace $ reverse name

  if null trimmedName
    then return "Stranger"
    else return trimmedName

newtype Greeter = Greeter String

greet :: Greeter -> IO ()
greet (Greeter name) = do
  putStrLn $ "Hello, " ++ name ++ "!"
  putStrLn "Greetings from Haskell!"

main :: IO ()
main = do
  name <- getUserName
  let greeter = Greeter name

  greet greeter

File: fortran/greeter/app/greeter.f90
program greeter
  implicit none

  character(len=256) :: username
  call input_handler(username)

  call greet(username)

contains

  subroutine input_handler(user_name)
    character(len=256), intent(out) :: user_name
    character(len=256) :: userinput
    character(len=256) :: name

    write(*, '(a)', advance='no') 'Enter your name: '
    read(*, '(a)') userinput

    name = trim(adjustl(userinput))

    if (len_trim(name) == 0) then
      name = 'Stranger'
    end if

    user_name = trim(name)
  end subroutine input_handler

  subroutine greet(greet_username)
    character(len=256), intent(in) :: greet_username

    write(*, "('Hello, ', a, '!')") trim(greet_username)
    write(*, "('Greetings from Fortran!')")
  end subroutine greet

end program greeter
